soroban-env-host/src/host/metered_xdr.rs

```diff
@@ -0,0 +1,28 @@
+use crate::{
+    budget::CostType,
+    xdr::{ReadXdr, WriteXdr},
+    Host, HostError,
+};
+
+impl Host {
+    pub(crate) fn metered_write_xdr(
+        &self,
+        obj: &impl WriteXdr,
+        buf: &mut Vec<u8>,
+    ) -> Result<(), HostError> {
+        obj.write_xdr(buf)
+            .map_err(|_| self.err_general("failed to write xdr"))?;
+        // Notes on metering": "write first charge later" means we could potentially underestimate
+        // the cost by the largest sized host object. Since we are bounding the memory limit of a
+        // host object, it is probably fine.
+        // Ideally, `charge` should go before `write_xdr`, which would require us to either 1.
+        // make serialization an iterative / chunked operation. Or 2. have a XDR method to
+        // calculate the serialized size. Both would require non-trivial XDR changes.
```

Comment:
It would nice to be able to do this, but realistically because XDR structures are nested, it would require mostly the same amount of work to calculate an accurate size as it would to just do the serialization. There a couple things we could do though, and one is very achievable today:

Put a `std::io::Write` implementation between the `buf` and the write_xdr function. The write_xdr function writes to a `std::io::Write`, which `Vec<u8>` just happens to implement. We can provide our own implementation that calls `charge_budget` as it goes along. You can do this today with just a few lines of code, and it would make the budget charging capability here much much safer.

Reply:
You mean reimplementing `std::io::write` for `Vec<u8>` to weave in budget charging? 
Or define a new type that contains `Vec<u8>` and implement `std::io::write` for it? 
The `HostError` in the return result might be an issue though.

Reply:
I mean something like this:

```rust
struct MeteredWrite<'a, W>
where
    W: Write,
{
    pub host: &'a Host,
    pub w: &'a mut W,
}

impl<'a, W> Write for MeteredWrite<'a, W>
where
    W: Write,
{
    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
        self.host
            .charge_budget(CostType::ValSer, buf.len() as u64)?;
        self.w.write(buf)
    }

    fn flush(&mut self) -> std::io::Result<()> {
        self.w.flush()
    }
}
```

You're that passing the `HostError` from `charge_budget` back up through the reader might be a problem. What I would do here is write a conversion for `HostError: Into<std::io::Error>` that results in the `HostError` being materialized back as it was in `HostError: From<std::io::Error>`.

Reply:
Brilliant solution, thanks @leighmcculloch !